### ISSUE TITLE
Scrollbar, file display misaligned

### ISSUE THREAD

**Issue by KimBruce** [gracelang/grace-web-editor#26]
When you click on a file after going elsewhere, the scrollbar is in the same place it was when you left the file before, but the content showing in the editor pane is from the beginning of the file.  Ideally the fix will be to adjust the content to match the scrollbar, but however it is fixed, the two should be consistent.

[There is a convenience to the current broken system.  Nudging the scrollbar by a tiny amount resets the editor pane to show you the place you were in the file.]

**Comment by Dmitri-2** [comment]
I think this issue occurs due to the editor simply swapping the content of the old file for the content of the new file. See [this line](https://github.com/gracelang/grace-web-editor/blob/pdx/scripts/editor.js#L301) in the editor code. If you have two long files in the editor, this can be seen if you scroll near the top on one of them and near the bottom on the other. Then switch from one to the other. The scrollbar and file should open in a similar place to where they were positioned in the old file. The ideal fix for this would be to have the editor create a new edit session for each file, and store the last known line position on the DOM file-tree. I am not quite sure how to fix this; I have tried adding `editor.getSession().setScrollTop(0);` after the content replacement command to have the editor scroll to the top on each file-open, but this appears to work only part of the time.

**Comment by apblack** [comment]
I really don't know how this works, but it seems that all he editor really needs to kno wis the character offset of the cursor.  Perhaps this could be saved as an attribute on whatever object is saved in the filestore, or as an attribute on the site-tree?

Kim's final comment, though — that nudging the scrollbar by a tiny amount resets the editor pane to the original place — makes me wonder if these attributes are in fact being restored correctly, but that there is a missing `refresh` command, or whatever command is used to synchronize the displayed state and the internal state.

**Comment by Dmitri-2** [comment]
This issue should be mostly fixed with commit  ebef736. This commit makes the editor track the cursor position and code folds for each file. However, the cursor position and scroll position are different. Should this issue be marked as closed? Or should the scroll position also be tracked?

**Comment by apblack** [comment]
Kim?  Are you happy with the new behaviour?

**Comment by KimBruce** [comment]
It seems like both the cursor position and the scroll position are tracked, even if they are different.  I.e., the cursor could be at the beginning of the file and the scroll bar at the bottom, and if you click on a different file and then return, the cursor will still be at the beginning and the scroll bar is at the bottom.  That seems fine to me.  If that is indeed the behavior (and others are happy with it) then we can close this.

**Comment by apblack** [comment]
The scroll position is not presently tracked.  The cursor position is now tracked, but it is possible that the cursor is not in the part of the file that is visible in the scrolling window.   The current behavior is that, when switching back to a previously-visited file, the editor will scroll so that the cursor is somewhere in the visible window.

Even if the scroll position were tracked, it turns out that there is a synchronization problem; the scroll-bar can't be reset until after the file has finished loading.  It seems to me that resetting the scroll position is a sufficiently minor issue that we should forget about it.  _However_, I don't do any serious programming in the web editor; if you think that the scroll position should also be maintained, we can certainly look at it.

**Comment by KimBruce** [comment]
Ah, I didn't play with it enough to get the exact behavior.  The version I have currently leaves the cursor where it was last, but does NOT scroll to make the cursor visible (unless I type something, of course).  I have the following version.
Grace Compiler v3813
IDE v148, commit 4041d6b

My own preference would be to retain the scroll bar position, as I'm usually bouncing back and forth looking at code, without necessarily editing it.  However, it is not a high enough priority to put a lot of effort into.  The most important thing is that the scroll bar now seems to accurately tell you where you are in the file when you return.

**Comment by apblack** [comment]
I played a bit more (in IDE v152, commit baf4173), and am seeing some broken behavior.   If I 

 1 . put the cursor at the top of a long file, F
 2. scroll to the bottom of F
 3. switch to another file G
 4. switch back to F
the cursor will be _one line higher_ than where I left it (but at the correct column position).

I'm guessing that this is an off by one error generated because ACE numbers the first line 0 internally.

I also, once, saw the scroll bar in a position that didn't correspond to the actual scrolling position.  That is, the scroll bar said that I was viewing the bottom of the file, whereas in fact I was viewing the top.  I couldn't reproduce this, though.  Nudging the scrollbar a tad fixed it.

**Comment by Dmitri-2** [comment]
I'll make sure to fix the off by one error. The cursor position values that cause this seem to be related to the way Ace sets and returns the cursor values. When retrieving the cursor position values, Ace provides the actual line numbers and columns, starting from "1." When setting the cursor position, however, Ace starts from 0. 

As for the scrolling problems, I was able to reproduce the issue of the scroll bar being in the wrong place. This only happens when the cursor is within the first 25 lines of a long file, and the scroll bar is elsewhere. I'll look into keeping track of the scroll position. Hopefully that will fix this issue.

**Comment by Dmitri-2** [comment]
This issue should be completely resolved with commit f4f1225. This commit fixes the off-by-one issue and has the editor keep track of the scrollbar position. This has also somehow fixed the issue where the editor would not scroll to the current cursor position on a page reload. Now, however, as the scrollbar is set separately, this is not a problem. I suspect the scroll position may be tracked in a different way, so it is set later in the editor load process. 

To recap, starting with IDE v154, the following are properly tracked and restored on a file open: 
1. The cursor position 
2. The scrollbar position 
3. All of the code folds in the file

**Comment by apblack** [comment]
> When retrieving the cursor position values, Ace provides the actual line numbers and columns, starting from "1." When setting the cursor position, however, Ace starts from 0.

Isn't consistency wonderful!
